### DISCUSSION TITLE
Scaling FrankenPHP - more workers vs multiple load-balanced instances?

### DISCUSSION THREAD

**Discussion by binaryfire** [php/frankenphp#527] [Q&A]
Hi all

Would increasing the FrankenPHP worker count in Laravel Octane increase req/s by (approximately) the same amount as running multiple load-balanced instances? For example, would would a single `--workers=8` instance be roughly equivalent to two `--workers=4` instances behind a load-balancer?

The reason I ask is that wasn't the case when I experimented with Octane running Swoole. Two load-balanced instances with 4 workers each were able to handle _significantly_ more req/s than a single instance with 8 workers. Which I found quite confusing...

This was also reflected in another user's benchmarks with Octane / Swoole: https://medium.com/@andrew_10845/laravel-octane-benchmarked-e430a43c6b33. You can see from the results that there was virtually no improvement in going from 4 to 8 workers.

**Comment by nickchomey** [comment]
I really have no idea, but my guess is that the issue had more to do with some other factor (ram, network bandwidth etc...).

Also the article you linked to is surely suffering from the fact that they used 1 vcpu for the many workers. Id want to see them do the same test with 4, 8, 20, 50 vcpus to see if the issue is cpus or worker threads.

**Reply by binaryfire** [reply]
Ah I hadn't noticed they were only using 1 core in that article. That would explain their results.

My testing was on a dedicated machine with 32 physical cores and 128gb RAM. The machine was definitely set up properly - there's some other factor that prevents the app scaling 1:1 (single instance w/ increased workers vs separate instances). At least as far as Octane + Swoole goes.

**Comment by withinboredom** [comment]
> Two load-balanced instances with 4 workers each were able to handle significantly more req/s than a single instance with 8 workers.

Yes, this makes sense. A single instance (if using https, especially), has to keep track of a lot of things and maintain each thread. Multiple instances can handle all of this independently. 

I've also noticed that FEWER workers tend to perform better than MORE workers, but I haven't tested it scientifically yet. I imagine there is a point where the overhead in maintaining the workers causes diminishing returns per worker.

**Reply by binaryfire** [reply]
@withinboredom Yeah that makes sense. I wasn't using https but the difference was still noticeable, especially when the server isn't a bottleneck. I'm using NGINX for ingress and just switched to the FrankenPHP Docker image so running multiple instances will be pretty straightforward. Seems like the best way to get the most out of bare metal machines. 

Very interested to see your test results when you publish them!
